src/scanners/javascript.js

```diff
@@ -78,83 +81,152 @@ export default class JavaScriptScanner {
+      // need to handle them specially.
+      if (message.fatal === true) {
+        // eslint-disable-next-line no-param-reassign
+        message.message = _messages.JS_SYNTAX_ERROR.code;
+      }
+
+      if (typeof message.message === 'undefined') {
+        throw new Error(
+          oneLine`JS rules must pass a valid message as
+          the second argument to context.report()`);
+      }
+
+      // Fallback to looking up the message object by the message
+      let code = message.message;
+      let shortDescription;
+      let description;
+
+      // Support 3rd party eslint rules that don't have our internal
+      // message structure and allow us to optionally overwrite
+      // their `message` and `description`.
+      if (Object.prototype.hasOwnProperty.call(_messages, code)) {
+        ({
           message: shortDescription,
-          sourceCode: message.source,
-          type: ESLINT_TYPES[message.severity],
-        });
+          description,
+        } = _messages[code]);
+      } else if (Object.prototype.hasOwnProperty.call(
+        messages.ESLINT_OVERWRITE_MESSAGE, message.ruleId)) {
+        const overwrites = messages.ESLINT_OVERWRITE_MESSAGE[message.ruleId];
+        shortDescription = overwrites.message || message.message;
+        description = overwrites.description || message.description;
+
+        if (overwrites.code) {
+          ({ code } = overwrites);
+        }
+      } else {
+        shortDescription = code;
+        description = null;
+      }
+
+      this.linterMessages.push({
+        code,
+        column: message.column,
+        description,
+        file: this.filename,
+        line: message.line,
+        message: shortDescription,
+        sourceCode: message.source,
+        type: ESLINT_TYPES[message.severity],
       });
     });
+
     return {
       linterMessages: this.linterMessages,
       scannedFiles: this.scannedFiles,
     };
   }
+
+  /*
+    Analyze the source-code by either looking into `javaScriptEs6Modules`
+    or by parsing the source code manually and check for import/export syntax
+    errors.
+
+    This set's `config.parserOptions.sourceType` correctly.
+  */
+  verifyParserOptionsAndGetSourceCode(config) {
+    if (this.javaScriptEs6Modules.includes(this.filename)) {
+      // eslint-disable-next-line no-param-reassign
+      config.parserOptions.sourceType = 'module';
+    }
+
+    // Default options taken from eslint/lib/linter:parse
+    // eslint-disable-next-line no-param-reassign
+    config.parserOptions = Object.assign({}, config.parserOptions, {
+      loc: true,
+      range: true,
+      raw: true,
+      tokens: true,
+      comment: true,
+      eslintVisitorKeys: true,
+      eslintScopeManager: true,
+      filePath: config.filename,
+    });
+
+    let ast;
+
+    try {
+      ast = espree.parse(this.code, config.parserOptions);
+    } catch (exc) {
+      if (exc.message.startsWith("'import' and 'export'")) {
```

Comment:
Having to resort to check the error message makes me sad :-( 
(and I'm sure that it makes you sad as well ;-))
and it also feels a bit fragile.

I suspect that this is also going to parse twice any js file that uses the import/export syntax in an invalid way because it also catch `"'import' and 'export' may only appear at the top level"` (instead of only catching `"'import' and 'export' may appear only with 'sourceType: module'"`).

If we can't avoid to match the error message, I would probably prefer if we check if the error `endsWith` `"may appear only with 'sourceType: module'"`.

Given that we are parsing the source, why we can't do the opposite?
- parse the source with parserOptions.sourceType = 'module'
- look into the ast for "top level import or export" statements:
  if there are => eslint should use `sourceType: "module"` for this file
  if there aren't => eslint should use `sourceType: "script"` for this file\

how that sounds to you?

Reply:
Hmm, let me try that for a moment. This might work, given that this would only parse it and avoid any rules being executed upfront and issue wrong warnings.

Reply:
So this works more or less… except for scenarios where `acorn` raises issues because it does indeed verify `eval` isn't being used in a `strict` context which happens when we parse as `sourceType: module`.

I am now parsing with `sourceType: module` by default and falling back to `script` which seems reasonable. I don't think it matters much whether we do it this way or the other way around.